port/stmicro/stm32/src/hals/STM32F103/uart.zig

```diff
@@ -0,0 +1,345 @@
+                            return error.Timeout;
+                        }
+                    }
+                }
+                regs.DR.raw = @intCast(byte);
+            }
+        }
+    }
+
+    pub fn readv_blocking(uart: UART, buffers: []const []u8, timeout: ?Timeout) ReceiveError!void {
+        const regs = uart.get_regs();
+        for (buffers) |buf| {
+            for (buf) |*bytes| {
+                while (!uart.is_readable()) {
+                    if (timeout) |check| {
+                        if (check.check_timeout()) {
+                            return error.Timeout;
+                        }
+                    }
+                }
+                const SR = regs.SR.read();
+
+                if (SR.ORE != 0) {
+                    return error.OverrunError;
+                } else if (SR.NE != 0) {
+                    return error.NoiseError;
+                } else if (SR.FE != 0) {
+                    return error.FramingError;
+                } else if (SR.PE != 0) {
+                    return error.ParityError;
+                }
+                const rx = regs.DR.raw;
+
+                bytes.* = @intCast(0xFF & rx);
+            }
+        }
+    }
+
+    pub fn get_errors(uart: UART) ErrorStates {
+        const regs = uart.get_regs();
+        const read_val = regs.SR.read();
+        return .{
+            .overrun_error = read_val.ORE == 1,
+            .noise_error = read_val.NE == 1,
+            .parity_error = read_val.PE == 1,
+            .framing_error = read_val.FE == 1,
+        };
+    }
+
+    pub inline fn clear_errors(uart: UART) void {
+        const regs = uart.get_regs();
+        std.mem.doNotOptimizeAway(regs.SR.raw);
+        std.mem.doNotOptimizeAway(regs.DR.raw);
+    }
+
+    pub fn write_blocking(uart: UART, data: []const u8, timeout: ?Timeout) TransmitError!void {
+        try uart.writev_blocking(&.{data}, timeout);
+    }
+
+    pub fn read_blocking(uart: UART, data: []u8, timeout: ?Timeout) ReceiveError!void {
+        try uart.readv_blocking(&.{data}, timeout);
+    }
+
+    pub fn writer(uart: UART) Writer {
+        return .{ .context = uart };
+    }
+
+    pub fn reader(uart: UART) Reader {
+        return .{ .context = uart };
+    }
+    fn generic_writer_fn(uart: UART, buffer: []const u8) TransmitError!usize {
+        try uart.write_blocking(buffer, null);
+        return buffer.len;
+    }
+
+    fn generic_reader_fn(uart: UART, buffer: []u8) ReceiveError!usize {
+        try uart.read_blocking(buffer, null);
+        return buffer.len;
+    }
+};
+
+var uart_logger: ?UART.Writer = null;
+
+/// Set a specific uart instance to be used for logging.
+///
+/// Allows system logging over uart via:
+/// pub const microzig_options = .{
+///     .logFn = hal.uart.logFn,
+/// };
+pub fn init_logger(uart: UART) void {
+    uart_logger = uart.writer();
+    uart_logger.?.writeAll("\r\n================ STARTING NEW LOGGER ================\r\n") catch {};
```

Comment:
this'll crash if we call it after `deinit_logger`. We should guard it like we do in `logFn`